Added typescript definitions.

geon opened the pull request:
I have gone through the existing unofficial typings file, and cleaned it up, modernized it, fixed a few errors and integrated it to work with the npm publishing script.

I also added most of the missing function definitions, including all of v3.

As a side effect, I discovered a few missing JSDoc definitions and fixed them.

I'm sure there are still errors in this definition, but it does compile.

geon commented:
There are a couple of functions that lacks documentation for one of the parameters. In particular, these overloaded functions:

```
createProgram
createProgramInfo
```

If someone can fix the documentation, as discussed here: https://github.com/greggman/twgl.js/issues/88

greggman commented:
looking at the types it seems like there's a bunch of issues with `ArrayBuffer` vs `ArrayBufferView`.

An `ArrayBuffer` is an untyped buffer of data you can not access the contents of directly. An `ArrayBufferView` is typed view into an `ArrayBuffer`.  

`ArrayBufferViews` are `Int8Array`, `Uint8Array`, `Int16Array`, `Uint16Array` `Int32Array`, `Uint32Array`, `Float32Array`, `Float64Array`.  

`ArrayBufferViews` contain a reference to the `ArrayBuffer` they are viewing

If I'm reading `index.d.ts` correctly it seems like many of the places that say `ArrayBuffer` should say `ArrayBufferView` and then `ArrayBufferView` probably needs to be defined as one of the various views (unless that's already defined for typescript somewhere)

geon commented:
Ah! I noticed that in one of the function definitions I added from scratch, and added a union for it. I'll go through them and make sure it gets used where it is needed.

Tyshark9 commented:
@geon I love you! 👍

NikitaIT commented:
Add lint for use `fun(): void` instead of `fun()` or use `fun()` instead of `fun(): void`.
Do not mix these two ways.
And
```
export function createUniformSetters(gl: WebGL2RenderingContext, program: WebGLProgram): ...;
```
instead of
```
export function createUniformSetters(program: WebGLProgram): ...
```

Will be better to quickly obtain not finalized definitions, and then finish, than chasing the ideal, but to roll out later, imho.

NikitaIT commented:
I did some work on types and want to offer it:
index.d.ts
```typescript
//help types, in typescript u can not put Base and Ext in same Type, but u can "&"
type SphereBuffersBase<TRestrictedAttribute extends string> = {
          [TKey in TRestrictedAttribute]: WebGLBuffer;
}
type SphereBuffersExt = { numElements: number; }
export type SphereBuffers<TRestrictedAttribute extends string> = 
          SphereBuffersExt & SphereBuffersBase<TRestrictedAttribute>;

export function createSphereBuffers<TFilds extends string>(...): SphereBuffers<TFilds>;
```
u-code.ts
```typescript
type TBuffers = 'position'|'normal'|'indices';
const buffers = twgl.primitives.createSphereBuffers<TBuffers>(gl, 10, 48, 24);
const attribs = {
      a_position: { buffer: buffers.position }, // ok
      a_normal: { buffer: buffers.normal }, // ok
      a_texcoord: { buffer: buffers.texcoord }, // TS2339: Property 'texcoord' does not exist
// on type 'SphereBuffers<"position" | "normal" | "indices">'.
};
```

pineapplemachine commented:
Any prospect of the repo gaining actually maintained TS definitions? Took me an additional 20 minutes of tinkering before tsc would accept the definitions, and I still can't say if everything is accurate and up-to-date. (Mostly I had to add missing `: void` return type declarations, and one completely missing return type for `createVertexArrayInfo`.)

greggman commented:
AFAIK the types are  wrong in the PR above and there's been no new PR to fix that.

Also the suggestions on `SphereBufferBase` etc I don't really understand. All the primitive functions return the same types of data.

If I get a .ts file that seems to work I'll be happy to take a look.

NikitaIT commented:
@pineapplemachine Most of the functions without `void` do not mean that they will `return void`.
I also did it first, but then I realized that I was wrong, just ignore this file for the linter.
u can copy SphereBuffersBase from [my try twgl repo](https://github.com/NikitaIT/WebGL-Typescript-Starter/blob/dev/src/types/twgl.js/index.d.ts)

pducks32 commented:
> no idea if the PR above works ... Maybe someone that knows typescript better can take a look?

They don't work great. I'm using this project in a Typescript project and have used it before from JS so I have some idea what needs tot be where and try my best to update them as I go. I will also look at the JSDoc conversion library, I know they exist I've just never used them myself.
